### DISCUSSION TITLE
Lanbon L8 freezes very often

### DISCUSSION THREAD

**Discussion by waltkind** [arendst/Tasmota#20442] [Support request] [answered]
I am very happy with the possibilities the Lanbon L8 has with HASPmota.
So I tried a lot and had no problem when the system froze during the tests. But now I came to the situation that I ran the system for longer without touching it and let it show some readings and waiting for interaction.
Now I see, that it freezes several times a day with no visible reason and can only be restarted by power cycling.

On the console screen there is nothing special to see. It only stops the state update every 300 seconds and is no more responsive.

Is this a known problem or do I have a defect system?
Is there a possibility to have more specific log or a watchdog script that restarts the system?

**Comment by sfromis** [comment]
Did you check if the available Heap size is stable, or counting down for some reason. Part of the STATE message at TelePeriod.

**Reply by waltkind** [reply]
Heap is stable between 40 and 44 all the time.
I don't know the margins what's a good size

**Reply by waltkind** [reply]
You are right. The heap was decreasing and it froze with heap 30. Uptime was 7:10:09 this time.
What can I do? What should I check?
The load average is constant at 99, the file system is free 4.204 MB of 4.416 MB

**Reply by sfromis** [reply]
Heap 40 was already on the low end, and dwindling memory available generally indicates a memory leak, needing debugging to work out where it gets lost.

With `SetOption130`, the console messages include more info about heap size and fragmentation with each time stamp, which can be useful to get closer.

The "LoadAVG" (which is not a load average) is rarely useful to look at.

**Reply by waltkind** [reply]
What is a reasonable heap size range?
I now deleted page 2 and 3, only page 1 is active and I start with heap 69 but it is decreasing.
SetOption130 is ON but I see no difference in logging

```berry
10:45:36.356-064/00 MQT: tasmota/lanbon_1/STATE = {"Time":"2024-01-10T10:45:36","Uptime":"0T00:20:10","UptimeSec":1210,"Heap":63,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":1,"Berry":{"HeapUsed":32,"Objects":588},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"ON","Dimmer":8,"Color":"0D1411","HSBColor":"156,36,8","Channel":[5,8,6],"Scheme":0,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"xxxx","BSSId":"98:9B:CB:9F:40:02","Channel":6,"Mode":"11n","RSSI":58,"Signal":-71,"LinkCount":1,"Downtime":"0T00:00:03"}}
``` 
Where can I start debugging?

**Reply by sfromis** [reply]
You **do** see a logging difference with the time stamp using that option.
> 10:45:36.356-**064/00**

That's the info about heap size/fragmentation, and seeing changes in these values might help get closer to where it happens, as this info is with all logged messages, not just the telemetry period messages.

In the STATE message, `"Berry":{"HeapUsed":32,"Objects":588}` also tells about the memory situation, and a growing trend here would be interesting.

I don't really have a "lower bound" for available heap, but down to 40 KB is where I'd start to look for why it got that low, when it is ESP32.

**Reply by waltkind** [reply]
"Berry":{"HeapUsed":32,"Objects":588} is stable and has been for days with 39 by 714 objects.

**Comment by Jason2866** [comment]
Does the devices has PSRAM? Without you won't get happy using HASPmota 
Graphic screens eat up a lot of RAM. Without PSRAM only very basic graphics can be done.
You can see if and how much PSRAM the device has in the Tasmota Information Web page

**Reply by sfromis** [reply]
This thread by an usually reliable source does show the device having an ESP32-WROVER-B module, which should mean PSRAM included.
https://blakadder.com/lanbon-L8-custom-firmware/
Som devices with an older ESP32 may need a build with board esp32-fix for Tasmota to detect the PSRAM, in this case there should be a message right at the start of the boot process.

To see these messages, restart from the console and push refresh a bit later to avoid the start of the log already getting lost by the time the console shows up by itself.

**Reply by waltkind** [reply]
It has no PSRAM. In another thread here I was told that I can use the Lanbon L8 without PSRAM and HASPmota.
I have the device since a month but I don't know how old it really is.
Is there a build ready with the fix?

```berry
00:00:00.003-204/00 HDW: ESP32-D0WD v1.0 (PSRAM disabled)
00:00:00.003-204/00 HDW: PSRAM is disabled, requires specific compilation on this hardware (see doc)
00:00:00.222-202/00 UFS: FlashFS mounted with 4204 kB free
00:00:00.238-202/00 CFG: Loaded from File, Count 520
00:00:00.247-202/00 QPC: Count 1
00:00:00.251-202/00 SPI: Hardware using GPIO19(CLK), GPIO23(MOSI) and GPIO25(MISO)
00:00:00.253-201/00 I2C: Bus1 using GPIO00(SCL) and GPIO04(SDA)
00:00:00.274-201/00 TFS: File 'mcp23x.dat' not found
00:00:00.432-193/00 BRY: Berry initialized, RAM used 5695 bytes
00:00:00.510-191/00 I2C: FT5206 found at 0x38
00:00:00.511-191/00 TI: FT5206
00:00:01.195-190/00 DSP: Configured display 'ST7789'
00:00:01.225-190/00 Project tasmota - Tasmota Version 13.3.0(lvgl-haspmota)-2_0_14(2023-12-12T14:31:28)
00:00:01.234-183/00 LVG: LVGL initialized
00:00:01.576-177/00 TAP: Loaded Tasmota App 'haspmota_widgets.tapp'
00:00:05.903-122/00 BRY: Successfully loaded 'autoexec.be'
00:00:09.277-084/00 WIF: Connecting to AP1 xxx Channel 6 BSSId 98:9B:CB:9F:40:02 in mode 11n as lanbon-1-0656...
00:00:11.295-077/00 WIF: Connected
```

**Reply by sfromis** [reply, marked as answer]
It likely **has** PSRAM, but with a rather old ESP32 it needs a build to support this hardware, as indicated in the console messages.
```
00:00:00.003-204/00 HDW: ESP32-D0WD v1.0 (PSRAM disabled)
00:00:00.003-204/00 HDW: PSRAM is disabled, requires specific compilation on this hardware (see doc)
```

I did not find where in the docs you're supposed to see, but what's needed is a build with board esp32-fix.
As an example, this happens for the webcam build (don't try using that build; no display support).
https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_env32.ini#L42-L50

When it comes to the LVGL build for displays on an ESP32 without working PSRAM, this is in the zone of being borderline, meaning that for a simple setup it may or may not work, and therefore not recommended, unless you experience no trouble. But you do :wink: 

To be clear, I'm not making a declaration that this is definitely the cause, more like doing general observations about the situation around PSRAM and support for displays.

**Reply by waltkind** [reply]
I fear that there is the problem. 
I used the desription of https://templates.blakadder.com/lanbon_L8-HS.html for flashing with the recommended tasmota32-lvgl.factory.bin. 
Building my own binary with this fix seems to be beyond my knowledge, but 'with a little help from my friends' there might be a way. I have no experience what is necessary and how to do it. 
Can you recommend a 'How-To' on this?

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/Compile-your-build/#defining-multiple-custom-firmwares shows how you can create "alternative binaries", in the same way as the webcam setup I linked.

**Reply by Jason2866** [reply]
In general it is a shame from the manufacturer to mount very very old esp32 chips. rev 1.0 is not produced since years! Maybe the use recycled components. Would not by this device.
Or you have got a faked device from the seller

**Reply by waltkind** [reply]
I don't know whether it is old hardware, but the price is perfect. 45.- EUR for a device that can be used as home automation center. 
The start is a bit bumpy but with the excellent help here, I am confident to get my needs solved.
I ordered a second device and will look if it is the same old esp32 chip.

**Comment by waltkind** [comment]
With your hints I compiled a new built with Gitpod and it looks quite promissing concerning the PSRAM, but is has another fault.
See the starting log. 
```berry
00:00:00.002-225/53 HDW: ESP32-D0WD v1.0 (PSRAM)
00:00:00.240-223/52 UFS: FlashFS mounted with 4204 kB free
00:00:00.259-223/52 CFG: Loaded from File, Count 524
00:00:00.267-223/52 QPC: Count 1
00:00:00.272-222/52 SPI: Hardware using GPIO19(CLK), GPIO23(MOSI) and GPIO25(MISO)
00:00:00.274-222/52 I2C: Bus1 using GPIO00(SCL) and GPIO04(SDA)
00:00:00.456-221/52 BRY: Berry initialized, RAM used 3961 bytes
00:00:00.543-220/52 Project tasmota - Tasmota Version 13.3.0(LVGLFIX)-2_0_14(2024-01-10T16:19:49)
00:00:00.853-220/52 TAP: Loaded Tasmota App 'haspmota_widgets.tapp'
00:00:01.372-220/52 BRY: failed to run compiled code 'syntax_error' - /haspmota_widgets.tapp#/lv_tasmota_log.be:3: 'lv' undeclared (first use in this function)
00:00:01.551-215/50 BRY: failed to run compiled code 'import_error' - module 'haspmota' not found
00:00:02.639-176/39 WIF: Connecting to AP1 xxx Channel 6 BSSId 98:9B:CB:9F:40:02 in mode 11n as lanbon-1-0656...
00:00:03.280-176/39 WIF: Connected
17:37:47.053-171/37 HTP: Web server active on lanbon-1-0656 with IP address 192.168.17.71
17:37:47.866-171/38 MQT: Attempting connection...
17:37:49.845-169/37 MQT: Connected
``` 

In platformio_tasmota_cenv.ini I used:
```berry
[env:tasmota32-lvglfix]
extends                 = env:tasmota32-lvgl
board                   = esp32-fix
board_build.f_cpu       = 240000000L
build_flags             = ${env:tasmota32_base.build_flags} -DFIRMWARE_TASMOTA-LVGLFIX
```
There is somewhere something missing. The screen is only white.

**Reply by waltkind** [reply]
I changed the build_flag to
 `build_flags             = ${env:tasmota32-lvgl.build_flags} -DFIRMWARE_TASMOTA-LVGLFIX`
and now it's OK.

**Reply by sfromis** [reply]
Your "new" build flag of `FIRMWARE_TASMOTA-LVGLFIX` does not make sense (is being used nowhere), and therefore you lose having LVGL included. Better remove both last two lines which you do not need to override when extending.

Note that when using HASPmota, you are **way** outside of the template suggestion for making the device work with display and without needing to use PSRAM. Nothing "wrong" about that, just to explain why you did not have luck with:
> In another thread here I was told that I can use the Lanbon L8 without PSRAM and HASPmota.

**Reply by ameeuw** [reply]
Thank you so much for this discussion. Helped me a lot as I was puzzled over the exact same behaviour of my Lanbon L8 over the last days.

**Comment by Jason2866** [comment]
This should be enough 
```
[env:tasmota32-lvglfix]
extends                 = env:tasmota32-lvgl
board                   = esp32-fix
```

**Reply by waltkind** [reply]
With the new build the device is now running more than three days stable with constant heap of 144 by 750 objects.
So I think I can mark that thread as answered and thank you for your help.

PS: The second device that I ordered has the same ESP v1.0 chip.
